### DISCUSSION TITLE
deno.exit() with Oak lib issue

### DISCUSSION THREAD

**Discussion by FrageDev** [denoland/deno#9382] [Q&A]
Hi,

I setup a web server with Oak and I have a specific route where the route return a JSON then close the socket and close the process.

My route is : 
```
.get('/api/reboot', async (ctx) => {
        await modules.reboot(ctx);
        await closeSocket();
})
```

The modules.reboot return the JSON with a "4" in the console after the ctx.response.body.
The function closeSocket() : 
```
async function closeSocket() {
    try {
    modules.logger.warning(`1`);
    await delay(5000);
    modules.logger.warning(`2`);
    controller.abort();
    await delay(5000);
    modules.logger.warning(`3`);
    Deno.exit();}
    catch (error){
        console.log(error)
    }
}
```
I added some delay to do some experimentation. Here the console log of the app : 

```
Wed Jan 06 2021 17:02:58 GMT+0100 (Romance Standard Time) | INFO         | Listening on https://localhost:8443
Wed Jan 06 2021 17:02:58 GMT+0100 (Romance Standard Time) | INFO         | GET https://localhost:8443/api/reboot from 127.0.0.1
Wed Jan 06 2021 17:02:58 GMT+0100 (Romance Standard Time) | INFO         | Reboot command recieved.
Wed Jan 06 2021 17:02:58 GMT+0100 (Romance Standard Time) | WARNING      | 4
Wed Jan 06 2021 17:02:58 GMT+0100 (Romance Standard Time) | WARNING      | 1
Wed Jan 06 2021 17:03:03 GMT+0100 (Romance Standard Time) | WARNING      | 2
Wed Jan 06 2021 17:03:08 GMT+0100 (Romance Standard Time) | WARNING      | 3
```

Here we can see the response body is supposed  to be sent and now proceed to kill the server listen and then kill the process.

The issue is I don't receive the JSON from the request. If I'm not wrong when I open the console from Chrome I can see the request, I can see the request header and also a response header but no body.

Request Headers : 
```
GET /api/reboot HTTP/1.1
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.101 Safari/537.36
```

Response Headers : 
```
HTTP/1.1 200 OK
content-length: 16
content-type: application/json; charset=utf-8
```

I don't know if I made any mistakes or deno issue or oak issue.

**Comment by Seally** [comment]
`Deno.exit()` as far as I know is a thin wrapper around the Rust exit function. Once that runs everything else that's running -- clean up code, incomplete asynchronous tasks (including the one handling your request) _etc._ -- is pretty much done for. If you need to gracefully shut down the server, better to read this instead: https://github.com/oakserver/oak#closing-the-server

Node has a similar issue with `process.exit()` which was why it has `process.exitCode` in case you want to set a non-zero exit code but not immediately exit the program. It's not Deno, but I think the explanation for the issue still somewhat applies: https://nodejs.org/api/process.html#process_process_exit_code

**Comment by FrageDev** [comment]
> `Deno.exit()` as far as I know is a thin wrapper around the Rust exit function. Once that runs everything else that's running -- clean up code, incomplete asynchronous tasks (including the one handling your request) _etc._ -- is pretty much done for. If you need to gracefully shut down the server, better to read this instead: https://github.com/oakserver/oak#closing-the-server
> 
> Node has a similar issue with `process.exit()` which was why it has `process.exitCode` in case you want to set a non-zero exit code but not immediately exit the program. It's not Deno, but I think the explanation for the issue still somewhat applies: https://nodejs.org/api/process.html#process_process_exit_code

@Seally  Thanks for your answer.

The problem is I put 5 secondes delay before exit(), so I should receive the response before because if I execute the same code without the exit I got the response.body in few ms.

For the oak#closing-the-server I do it but the process is still alive, that the reason why in my closeSocket() function I close the web server listening then close the deno process.

```
controller.abort();
Deno.exit();
```

**Comment by Seally** [comment]
`Deno.exit()` isn't required to exit the Deno process. If it is, a file with just `console.log("abc")` will never exit.

In Node, the process exits naturally if (quoting the docs) "there is no additional work pending in the event loop." I'm not sure how Deno handles it, but I expect it to be similar. Instead of calling `Deno.exit()`, I think it's better to structure your program in a way that closing the server also exits the program. I won't normally use `Deno.exit()` unless I need to set an exit code or I'm okay with the "burn everything down and run" kind of exit for that particular situation. 

What happens if you simply don't have `Deno.exit()`? I mean, I wrote this and it seems to exit, no unsafe kill-switch function required:

```typescript
import { Application, Router } from "https://deno.land/x/oak/mod.ts";

const app = new Application();

const controller = new AbortController();
const { signal } = controller;

const router = new Router();

router.get("/reboot", () => {
    controller.abort();
});

app.use(router.routes());
app.use(router.allowedMethods());

await app.listen({ port: 8000, signal });
```

**Comment by FrageDev** [comment]
@Seally This is what I'm going to do, I will just kill the websocket and not the process cause after my function the PC is supposed to reboot so will kill deno process.

And the Deno.exit() seems to don't close the process (I still need to ctrl v on console to go out)

**Comment by CanRau** [comment]
I just published [an article](https://www.canrau.com/en/gracefully-close-sub-subprocess-using-signals-in-deno) explaining and showing in examples how I close my oak server which I started from my custom CLI using `Deno.run` using signals.
The original issue (now discussion) helped me a lot and I thought others might profit from what I've learned and how I've done it ☺️
